src/main/daml/Daml/Finance/Instrument/Swap/Fpml/Util.daml

```diff
@@ -615,30 +617,31 @@ calculateFloatingPaymentClaimsFromSwapStream floatingRateCalculation s
                 if pp.rateFixingDate < pp.calculationPeriod.adjustedEndDate
                 then error "fixing of notional must not occur before the final calculation date"
                 else
-                  when (TimeGte pp.rateFixingDate) $ scale notional $
-                  scale (convertImplicitDcfToActualDcf) $
-                  when (TimeGte p.adjustedEndDate) $ one currency
+                  scale (pp.notional * convertImplicitDcfToActualDcf) $ one currency
+              rateStart = ObserveAt floatingRateCalculation.floatingRateIndex
+                pp.calculationPeriod.adjustedStartDate
+              rateEnd = ObserveAt floatingRateCalculation.floatingRateIndex
+                pp.calculationPeriod.adjustedEndDate
               claim =
                 -- calculate a daily compounded reference rate using the corresponding index,
                 -- see page 36 here:
                 -- https://www.newyorkfed.org/medialibrary/Microsites/arrc/files/2021/users-guide-to-sofr2021-update.pdf
-                scaleByAssetPerformance pp.calculationPeriod.adjustedStartDate
-                  pp.calculationPeriod.adjustedEndDate rate endClaim
+                when (TimeGte p.adjustedEndDate) $ scale (rateEnd / rateStart - Const 1.0) endClaim
 
       -- Calculate a floating rate correction claim for a payment period
       -- (corresponding to the "-1" in a perf calculation)
       -- This will no longer be needed when improved compounding logic is supported in CC
       createCorrectionClaim p =
         give $ scale notional $ when (TimeGte p.adjustedEndDate) $ one currency
         where
-          (notional, _) = calculateNotionalAndFixings p s calculationScheduleStaticData fxRateId
-            useAdjustedDatesForDcf periodicSchedule
+          (Some notional, _) = calculateNotionalAndFixings p s calculationScheduleStaticData
+            fxRateId useAdjustedDatesForDcf periodicSchedule
 
       -- calculate floating rate spread claim for a payment period
       createSpreadClaim p =
         mconcat $ simpleInterestSpreadClaims
         where
-          (notional, paymentPeriodData) = calculateNotionalAndFixings p s
+          (Some notional, paymentPeriodData) = calculateNotionalAndFixings p s
```

Comment:
In what case would notional be returned as `None`? Currently the error message would be quite hard to read if notional is returned as `None`

Reply:
It feels that the `Some` is sort of redundant

Reply:
Some swaps have an FX adjusted notional, which means the notional is not guaranteed to be constant over the payment period. I have added an error message for this.

Reply:
Ok, so in that case you are explicitly looking for a constant notional over the payment period.